### ISSUE TITLE
Rainbow Grades generate broken report when extra credit category is added.

### ISSUE THREAD

**Issue by KCony** [Submitty/Submitty#9469]
**Describe the bug**
Rainbow Grades report (output.html) gets broken when an extra credit section is added to customization.json. See the screen shot. The sorting order is way off, there are no values in the "Overall" column, the "Perfect" row is way down in the middle of the list.

**Expected behavior**
Rainbow Grades report should not get broken when a valid extra credit section is added. If anything is wrong with customization.json, Rainbow Grades should fail to "make" and report an error at this time rather than building normally but silently producing a broken report.

**To Reproduce**
1. Start with the initial customization_old.json (attached)
2. Build Rainbow Grades and observe the normal appearance of the report (screenshot attached)
3. Now add the following section (see file customization.json):
```
    {
      "count": 1,
      "percent": 0.08,
      "type": "Participation",
      "ids": [
        {
          "max": 0.0,
          "id": "participation",
          "released": false,
          "scale_max" : 8.0,
          "extra_credit" : true
        }
      ]
    }
```
4. Build Rainbow Grades and open output.html.
5. Observe broken report (see the first screenshot)

**Configuration**
 - OS: Windows 11 Home Version 22H2 OS build 22621.1848 Experience Windows Feature Experience Pack 1000.22642.1000.0
 - Browser: Mozilla Firefox 114.0.2 (64-bit)

**Screenshots**
![image](https://github.com/Submitty/Submitty/assets/7108039/6c3693b0-5035-4111-85a6-44406aa21675)
![image](https://github.com/Submitty/Submitty/assets/7108039/27546f48-a47d-4b36-a2c8-06e7eec64db5)

**Additional context**
[RainbowData.zip](https://github.com/Submitty/Submitty/files/11849495/RainbowData.zip)

**Comment by KCony** [comment]
Thank you for looking into that, @ziesski ! For "extra_credit", I have been using it for ages, and didn't realize it was no longer necessary. I really wish Submitty documentation could be updated and expanded on the usage of extra credit options (e.g., an extra credit gradeable within a group of regular gradeables or an entire type of gradeables representing extra credit).

Anyways, I tried your suggestion, and with the following in my customization.json:

```
    {
      "count": 1,
      "percent": 0.08,
      "type": "Participation",
      "ids": [
        {
          "max": 0.0,
          "id": "participation",
          "released": true,
          "scale_max" : 8.0
        }
      ]
    }
```

the issue still exists:

![image](https://github.com/Submitty/Submitty/assets/7108039/e20e3c1c-dfa6-47bd-9307-f620e958be1d)

Please let me know if you want me to collect any additional information.

**Comment by bmcutler** [comment]
Hi @KCony 

I tried to reproduce this on my Mac, with your customization file, your course data and the latest Rainbow Grades.  I could not reproduce the problem.
Are you running this on Windows?  If so, maybe we have a C++ portability error.  Could you test this running it on Ubuntu (e.g., run RainbowGrades on Submitty).  If that's the case I'll ask @ziesski to test & debug on Windows (he uses Mac as his primary platform but has access to Windows).
Also, it probably won't fix it (we don't think we've changed any relevant code), but can you try using the latest RainbowGrades?

**Comment by KCony** [comment]
Hi @bmcutler ,

Sorry for the confusion, I just copied and pasted my Windows/Firefox version info to the bug report but I actually run Rainbow Grades on Kubuntu and then only use Windows and Firefox to view the output.html. My Kubuntu is under VirtualBox, if it matters, so it is not the Submitty machine. Do you want me to run Rainbow specifically on Submitty and see what happens?

I will try the latest RainbowGrades later tonight or tomorrow and update you on whether it changed anything.

**Comment by ziesski** [comment]
I just tested on Windows 10, VirtualBox 7, Chrome - was not able to reproduce the issue. @KCony Barb is out of town, may not get back to you soon, I think it is reasonable to run Rainbow specifically on Submitty and see the behavior. Moreover, your initial suggestion is indeed valid, Rainbow Grade should rather fail to build than to build broken. We will develop an additional check layer for it after some discussion.

**Comment by KCony** [comment]
OK, I figured out what is going on. I have a bunch of patches that I wrote (or that Wes wrote earlier) to fix various issues in students.cpp. Apparently, some of these patches are broken by the newest changes in the RainbowGrades code. I would just keep using the older Rainbow version if I could but unfortunately there were some breaking changes to *.json files in the past year or so (adding new sections) that broke old Rainbow Grades, so I was forced to upgrade to the latest version.

I thought that I submitted all my patches in the PR long time ago but I either I didn't (or maybe not all of them) or the PR(s) never got approved, so my changes never made it to the main Rainbow code. These patches fix some bugs that are very important for me. Particularly, the stock version of Rainbow has bugs in computing percentages of gradeables in the group (by ignoring the relative weight of each gradeable). It also has a nasty recent addition of hardcoded failing the students which is not applicable to my class but since it's hardcoded, I have to patch the cpp. In addition, it has bugs with optional extra credit gradeables.

Anyhow, I will review my patches, so that they continue to work correctly. I'll probably have to get rid of some of the stock Rainbow code that gets in my way but since such changes are local, they shouldn't affect anyone else, unless some other instructors notice the same bugs that I just fixed locally with my own patches. :)

I guess this issue can be closed since other users will not be getting this broken report, and I will fix my patches locally to continue to work correctly. Sorry for the trouble and thank you for looking into this.
